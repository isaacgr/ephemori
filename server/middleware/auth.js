const {
  UnauthenticatedError,
  UserNotVerifiedError
} = require("../utils/errors");
const UserService = require("../db/User");
const SessionService = require("../db/Session");

class AuthService {
  constructor(db) {
    this.db = db;
  }

  get userService() {
    return this.UserService;
  }

  set userService(service) {
    this.UserService = service;
  }

  get sessionService() {
    return this.SessionService;
  }

  set sessionService(service) {
    this.SessionService = service;
  }

  async isAuthenticated(req, res, next) {
    this.userService = new UserService(this.db);
    this.sessionService = new SessionService(this.db);
    try {
      if (
        !req.session ||
        !req.session.user ||
        !req.cookies ||
        !req.cookies["connect.sid"]
      ) {
        throw new UnauthenticatedError("User is not authenticated");
      }
      await this.userService.getUserById(req.session.user.id);
      /**
       * TODO: Do i even need to do this? Does the middleware handle it?
       */
      // await this.sessionService.getSession(
      //   req.cookies["connect.sid"].split(":")[1].split(".")[0]
      // );
      next();
    } catch (e) {
      res.status(401).json({
        errorMessage: `${e.code} : ${e.message}`
      });
    }
  }

  async isUserVerified(req, res, next) {
    this.userService = new UserService(this.db);
    try {
      const verified = await this.userService.getUserVerfied(req.body.email);
      if (!verified) {
        throw new UserNotVerifiedError("User has not verified their account");
      }
      next();
    } catch (e) {
      res.status(401).json({
        errorMessage: `${e.code} : ${e.message}`
      });
    }
  }
}

module.exports = AuthService;
