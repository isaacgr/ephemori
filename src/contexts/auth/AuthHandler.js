import { HttpMethod, Endpoint } from "../../util/constants";
import { performApiRequest } from "../../util/functions";

export class AuthHandler {
  initialize(connectionSettings) {
    const { databaseURL } = connectionSettings;
    this.databaseURL = databaseURL;
    if (!(this instanceof AuthHandler)) {
      return new AuthHandler();
    } else {
      return this;
    }
  }

  async createUserWithEmailAndPassword(email, password) {
    return await performApiRequest(this, HttpMethod.POST, Endpoint.SIGN_UP, {
      email,
      password
    });
  }

  async signInWithEmailAndPassword(email, password) {
    return await performApiRequest(this, HttpMethod.POST, Endpoint.LOGIN, {
      email,
      password
    });
  }

  async checkAuthenticated() {
    return await performApiRequest(
      this,
      HttpMethod.GET,
      Endpoint.GET_ACCOUNT_INFO
    );
  }

  async verifySignupRequest(id, token) {
    return await performApiRequest(
      this,
      HttpMethod.GET,
      Endpoint.EMAIL_VERIFICATION,
      {
        id,
        token
      }
    );
  }

  async requestResetPassword(email) {
    return await performApiRequest(
      this,
      HttpMethod.POST,
      Endpoint.RESET_PASSWORD,
      { email }
    );
  }

  async resetPassword(id, token, password) {
    return await performApiRequest(
      this,
      HttpMethod.PATCH,
      Endpoint.RESET_PASSWORD,
      { id, token, password }
    );
  }

  async checkResetToken(id, token) {
    return await performApiRequest(
      this,
      HttpMethod.GET,
      Endpoint.RESET_VERIFICATION,
      {
        userId: id,
        token
      }
    );
  }

  async signOut() {
    return await performApiRequest(this, HttpMethod.GET, Endpoint.LOGOUT);
  }

  async setImportantDates(importantDates) {
    return await performApiRequest(
      this,
      HttpMethod.POST,
      Endpoint.SET_IMPORTANT_DATES,
      { importantDates }
    );
  }
}
