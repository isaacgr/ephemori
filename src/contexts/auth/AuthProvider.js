import React, { useContext, useState, useEffect } from "react";
import LoadingBar from "../../components/LoadingBar";
import { AuthHandler } from "./AuthHandler";

const AuthContext = React.createContext();

export const useAuth = () => {
  return useContext(AuthContext);
};

export const AuthProvider = ({ children }) => {
  const [currentUser, setCurrentUser] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");
  const authHandler = new AuthHandler();

  const auth = authHandler.initialize({
    databaseURL: `${process.env.API_SERVER_URL}`
  });

  useEffect(async () => {
    let isMounted = true;
    (async () => {
      try {
        if (isMounted) {
          setLoading(true);
          const res = await auth.checkAuthenticated();
          if (res.user) {
            setCurrentUser({
              ...res.user
            });
          } else {
            console.log("No user returned from checkAuthenticated");
          }
        }
      } catch (e) {
        if (isMounted) {
          setError(e.message);
        }
      } finally {
        if (isMounted) {
          setLoading(false);
        }
      }
    })();
    return () => {
      isMounted = false;
    };
  }, []);

  const signup = async (email, password, reRequest) => {
    try {
      const response = await auth.createUserWithEmailAndPassword(
        email,
        password,
        reRequest
      );
      setCurrentUser(response.user);
      return response;
    } catch (e) {
      throw e;
    }
  };

  const login = async (email, password) => {
    try {
      const response = await auth.signInWithEmailAndPassword(email, password);
      setCurrentUser(response.user);
      return response;
    } catch (e) {
      throw e;
    }
  };

  const logout = async () => {
    try {
      const response = await auth.signOut();
      return response;
    } catch (e) {
      throw e;
    } finally {
      setCurrentUser(null);
    }
  };

  const verifySignupRequest = async (id, token) => {
    try {
      const response = await auth.verifySignupRequest(id, token);
      return response;
    } catch (e) {
      throw e;
    }
  };

  const requestResetPassword = async (email) => {
    try {
      const response = await auth.requestResetPassword(email);
      return response;
    } catch (e) {
      throw e;
    }
  };

  const resetPassword = async (id, token, password) => {
    try {
      const response = await auth.resetPassword(id, token, password);
      return response;
    } catch (e) {
      throw e;
    }
  };

  const checkResetTokenValid = async (id, token) => {
    try {
      const response = await auth.checkResetToken(id, token);
      return response;
    } catch (e) {
      throw e;
    }
  };

  const value = {
    currentUser,
    error,
    login,
    signup,
    logout,
    verifySignupRequest,
    requestResetPassword,
    resetPassword,
    checkResetTokenValid
  };

  return (
    <AuthContext.Provider value={value}>
      {loading ? (
        <div className="container">
          <div className="content-block">
            <LoadingBar />
          </div>
        </div>
      ) : (
        children
      )}
    </AuthContext.Provider>
  );
};
