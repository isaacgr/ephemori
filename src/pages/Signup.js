import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import Paper from "@mui/material/Paper";
import Typography from "@mui/material/Typography";
import TextField from "@mui/material/TextField";
import FormControl from "@mui/material/FormControl";
import FormGroup from "@mui/material/FormGroup";
import Alert from "@mui/material/Alert";
import FormButton from "../components/FormButton";
import { useAuth } from "../contexts/auth/AuthProvider";
import FormControlLabel from "@mui/material/FormControlLabel";
import Checkbox from "@mui/material/Checkbox";
import Footer from "../components/Footer";

const Signup = () => {
  const [emailRef, setEmailRef] = useState("");
  const [passwordRef, setPasswordRef] = useState("");
  const [passwordConfirmRef, setPasswordConfirmRef] = useState("");
  const { signup } = useAuth();
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);
  const [signupResponse, setSignupResponse] = useState({});
  const [agree, setAgree] = useState(false);

  const handleResend = async (e) => {
    e.preventDefault();
    try {
      setError("");
      setLoading(true);
      const { verificationRequestSent, message } = await signup(
        emailRef,
        passwordRef
      );
      setSignupResponse(() => ({
        verificationRequestSent,
        message
      }));
    } catch (e) {
      setError(`Failed to send email: ${e.message}`);
    }
    setLoading(false);
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (!agree) {
      return setError(
        "You must agree to the terms and conditions and agree that you have read the privacy policy"
      );
    }
    if (passwordRef !== passwordConfirmRef) {
      return setError("Passwords do not match");
    }
    try {
      setError("");
      setLoading(true);
      const { success, message } = await signup(emailRef, passwordRef);
      setSignupResponse(() => ({
        success,
        message
      }));
    } catch (e) {
      setError(`Failed to create account: ${e.message}`);
    }
    setLoading(false);
  };

  return (
    <div className="container">
      <div className="content-block">
        <Paper className="auth-page auth-page__signup" variant="outlined">
          {signupResponse.success ? (
            <form onSubmit={handleResend}>
              <div className="content-block">
                <Alert severity="success">Email sent to {emailRef}</Alert>
                {error && <Alert severity="error">{error}</Alert>}
                <FormGroup className="content-block">
                  <FormButton
                    buttonText="Re-send Verification Email"
                    loading={loading}
                  />
                </FormGroup>
              </div>
            </form>
          ) : (
            <>
              <div className="content-block">
                <Typography variant="h4">Sign Up</Typography>
                {error && <Alert severity="error">{error}</Alert>}
              </div>
              <form onSubmit={handleSubmit}>
                <FormControl>
                  <FormGroup className="content-block">
                    <TextField
                      type="email"
                      required
                      label="Email"
                      variant="standard"
                      value={emailRef}
                      onChange={(e) => setEmailRef(e.target.value)}
                    />
                    <TextField
                      type="password"
                      required
                      label="Password"
                      variant="standard"
                      value={passwordRef}
                      onChange={(e) => setPasswordRef(e.target.value)}
                    />
                    <TextField
                      type="password"
                      required
                      label="Password Confirmation"
                      variant="standard"
                      value={passwordConfirmRef}
                      onChange={(e) => setPasswordConfirmRef(e.target.value)}
                    />
                  </FormGroup>
                  <FormGroup className="content-block">
                    <FormButton buttonText="Sign Up" loading={loading} />
                  </FormGroup>
                </FormControl>
              </form>
              <FormGroup>
                <FormControlLabel
                  control={
                    <Checkbox
                      color="secondary"
                      checked={agree}
                      onChange={(e) => {
                        setAgree(e.target.checked);
                      }}
                    />
                  }
                  label={`By checking this box you agree to the terms and conditions and agree that you have read and understood our privacy policy`}
                />
              </FormGroup>
            </>
          )}
        </Paper>
      </div>
      <div className="content-block">
        <Typography variant="subtitle1">
          Already have an account? <Link to="/login">Log In</Link>
        </Typography>
      </div>
    </div>
  );
};

export default Signup;
