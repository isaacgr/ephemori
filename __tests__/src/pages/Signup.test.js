/** @jest-environment jsdom */
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import * as Auth from "../../../src/contexts/auth/AuthProvider";
import renderer from "react-test-renderer";
import Signup from "../../../src/pages/Signup";

const user = {
  userId: 1,
  email: "testsuite@email.com",
  displayName: null,
  dateOfBirth: "1992-01-29",
  isUserSet: true
};

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate
}));

jest.spyOn(Auth, "useAuth").mockImplementation(() => {
  return {
    currentUser: user
  };
});

describe("<Signup/>", () => {
  it("should render correctly when loaded", () => {
    const login = renderer
      .create(
        <Router>
          <Signup />
        </Router>
      )
      .toJSON();
    expect(login).toMatchSnapshot();
  });
});
