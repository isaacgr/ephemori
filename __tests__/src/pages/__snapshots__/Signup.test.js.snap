// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> should render correctly when loaded 1`] = `
<div
  className="container"
>
  <div
    className="content-block"
  >
    <div
      className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded auth-page auth-page__signup css-172kgz0-MuiPaper-root"
    >
      <div
        className="content-block"
      >
        <h4
          className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          Sign Up
        </h4>
        
      </div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            className="MuiFormGroup-root content-block css-dmmspl-MuiFormGroup-root"
          >
            <div
              className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink={false}
              >
                Email
                <span
                  aria-hidden={true}
                  className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={true}
                  type="email"
                  value=""
                />
              </div>
            </div>
            <div
              className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink={false}
              >
                Password
                <span
                  aria-hidden={true}
                  className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={true}
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink={false}
              >
                Password Confirmation
                <span
                  aria-hidden={true}
                  className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={true}
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="MuiFormGroup-root content-block css-dmmspl-MuiFormGroup-root"
          >
            <div
              className="MuiBox-root css-1tfqlfq"
            >
              <button
                className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
                disabled={false}
                onBlur={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="submit"
              >
                Sign Up
              </button>
            </div>
          </div>
          <label
            className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              className="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1sg8fd1-MuiButtonBase-root-MuiCheckbox-root"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate={false}
                disabled={false}
                onChange={[Function]}
                type="checkbox"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              Re-send verification email
            </span>
          </label>
        </div>
      </form>
      <div
        className="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
      >
        <label
          className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            className="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1sg8fd1-MuiButtonBase-root-MuiCheckbox-root"
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <input
              checked={false}
              className="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate={false}
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            By checking this box you agree to the terms and conditions and agree that you have read and understood our privacy policy
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="content-block"
  >
    <h6
      className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
    >
      Already have an account? 
      <a
        href="/login"
        onClick={[Function]}
      >
        Log In
      </a>
    </h6>
  </div>
</div>
`;
